modules/fedimint-ln-client/src/lib.rs

```diff
@@ -319,42 +336,50 @@ impl LightningClientExt for Client {
 
         let refund_success = lightning.await_refund(operation_id);
 
-        Ok(Box::pin(stream! {
+        let stream = stream! {
             yield LnPayState::Created;
 
-            match tx_accepted_future.await {
-                Ok(()) => {
-                    yield LnPayState::Funded;
+            if tx_accepted_future.await.is_err() {
+                yield LnPayState::Canceled;
+                return;
+            }
 
-                    // await success or waiting for refund
-                    match payment_success.await {
-                        Ok(preimage) => {
-                            yield LnPayState::Success {preimage};
-                        }
-                        Err(LightningPayError::Refundable(refundable)) => {
-                            yield LnPayState::WaitingForRefund{ block_height: refundable };
-
-                            // in case of refund, await output of refund
-                            match refund_success.await {
-                                Ok(refund_txid) => {
-                                    yield LnPayState::Refunded { refund_txid };
-                                }
-                                Err(_) => {
-                                    yield LnPayState::Failed;
-                                }
+            yield LnPayState::Funded;
+
+            match payment_success.await {
+                Ok(preimage) => {
+                    if let Some(change) = change_outpoint {
+                        yield LnPayState::AwaitingChange;
+                        match self.await_primary_module_output_finalized(operation_id, change).await {
+                            Ok(_) => {}
+                            Err(_) => {
+                                yield LnPayState::Failed;
+                                return;
                             }
                         }
-                        Err(_) => {
-                            error!("Unexpected error state after funding transaction");
-                        }
                     }
 
-                },
-                Err(_) => {
-                    yield LnPayState::Canceled;
+                    yield LnPayState::Success {preimage};
+                    return;
                 }
+                Err(LightningPayError::Refundable(block_height, error)) => {
+                    yield LnPayState::WaitingForRefund{ block_height, gateway_error: error.clone() };
+
+                    if let Ok(refund_txid) = refund_success.await {
+                        // need to await primary module to get refund
+                        if self.await_primary_module_output_finalized(operation_id, OutPoint{ txid: refund_txid, out_idx: 0}).await.is_ok() {
+                            yield LnPayState::Refunded { gateway_error: error };
+                            return;
+                        }
```

Comment:
I wonder if we could get rid of the `LightningPayStates::Refunded` state now and instead only rely on `await_primary_module_output_finalized` to determine the high-level operation state (for the output to be finalized the tx has to be accepted). Could be a follow-up.